pkg/schedule/schedulers/balance_range.go

```diff
@@ -0,0 +1,816 @@
+			balancePersistFailedCounter.Inc()
+			return false
+		}
+		s.cleanJobStatus(cluster, job)
+		balanceRangeExpiredCounter.Inc()
+		log.Info("balance key range job finished due to timeout", zap.String("alias", job.Alias), zap.Uint64("job-id", job.JobID))
+	}
+
+	opInfluence := s.OpController.GetOpInfluence(cluster.GetBasicCluster(), operator.WithRangeOption(job.Ranges))
+	err := s.prepare(cluster, opInfluence, job)
+	if err != nil {
+		log.Warn("failed to prepare balance key range scheduler", errs.ZapError(err))
+		return false
+	}
+	if s.isBalanced() {
+		if err = s.conf.finish(index); err != nil {
+			balancePersistFailedCounter.Inc()
+			return false
+		}
+		s.cleanJobStatus(cluster, job)
+		balanceRangeBalancedCounter.Inc()
+		log.Info("balance key range job finished due to balanced", zap.String("alias", job.Alias), zap.Uint64("job-id", job.JobID))
+		return false
+	}
+	return true
+}
+
+func (s *balanceRangeScheduler) cleanJobStatus(cluster sche.SchedulerCluster, job *balanceRangeSchedulerJob) {
+	km := cluster.GetKeyRangeManager()
+	km.Delete(job.Ranges)
+	for storeID := range s.scoreMap {
+		storeStr := strconv.FormatUint(storeID, 10)
+		balanceRangeGauge.WithLabelValues(storeStr, "score").Set(0)
+		if _, ok := s.expectScoreMap[storeID]; ok {
+			balanceRangeGauge.WithLabelValues(storeStr, "expect").Set(0)
+		}
+	}
+}
+
+// BalanceRangeCreateOption is used to create a scheduler with an option.
+type BalanceRangeCreateOption func(s *balanceRangeScheduler)
+
+// newBalanceRangeScheduler creates a scheduler that tends to keep given peer rule on
+// special store balanced.
+func newBalanceRangeScheduler(opController *operator.Controller, conf *balanceRangeSchedulerConfig, options ...BalanceRangeCreateOption) Scheduler {
+	s := &balanceRangeScheduler{
+		BaseScheduler: NewBaseScheduler(opController, types.BalanceRangeScheduler, conf),
+		conf:          conf,
+		handler:       newBalanceRangeHandler(conf),
+	}
+	for _, option := range options {
+		option(s)
+	}
+
+	s.filters = []filter.Filter{
+		&filter.StoreStateFilter{ActionScope: s.GetName(), TransferLeader: true, MoveRegion: true, OperatorLevel: constant.Medium},
+		filter.NewSpecialUseFilter(s.GetName()),
+	}
+	s.filterCounter = filter.NewCounter(s.GetName())
+	return s
+}
+
+// Schedule schedules the balance key range operator.
+func (s *balanceRangeScheduler) Schedule(cluster sche.SchedulerCluster, _ bool) ([]*operator.Operator, []plan.Plan) {
+	balanceRangeCounter.Inc()
+	_, job := s.conf.peek()
+	if job == nil {
+		balanceRangeNoJobCounter.Inc()
+		return nil, nil
+	}
+	defer s.filterCounter.Flush()
+
+	faultStores := filter.SelectUnavailableTargetStores(s.stores, s.filters, cluster.GetSchedulerConfig(), nil, s.filterCounter)
+	sources := filter.SelectSourceStores(s.stores, s.filters, cluster.GetSchedulerConfig(), nil, s.filterCounter)
+	solver := s.solver
+	replicaFilter := filter.NewRegionReplicatedFilter(cluster)
+	baseRegionFilters := []filter.RegionFilter{
```

Comment:
Do we need to init filters according to the job rule?

Reply:
No need, it can move the leader if the rule is leader-scatter.